File: go/drift/snapshot_store.go
// SPDX-License-Identifier: AGPL-3.0-or-later

package drift

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	pg "github.com/dogukanturhal/keystone-sdk/go/sqlquote"
)

// snapshotTableDDL extends keystone_baselines (which only stores hashes)
// with a JSONB column for the canonical snapshot. The DriftController
// writes the snapshot when it accepts a baseline and reads it when
// drift is detected to compute structured findings.
//
// Co-located with the existing baseline table to avoid adding another
// per-schema bookkeeping table.
const snapshotTableDDL = `
ALTER TABLE %s.keystone_baselines
	ADD COLUMN IF NOT EXISTS snapshot JSONB;
`

// EnsureSnapshotColumn is a one-shot idempotent migration that adds the
// snapshot JSONB column to keystone_baselines. Called every reconcile
// — the IF NOT EXISTS makes it cheap.
func EnsureSnapshotColumn(ctx context.Context, pool *pgxpool.Pool, schema string) error {
	if err := pg.ValidateIdentifier("schema", schema); err != nil {
		return err
	}
	stmt := fmt.Sprintf(snapshotTableDDL, pg.QuoteIdentifier(schema))
	if _, err := pool.Exec(ctx, stmt); err != nil {
		return fmt.Errorf("alter keystone_baselines add snapshot: %w", err)
	}
	return nil
}

// ReadSnapshot returns the prior snapshot JSON, or nil if none recorded.
func ReadSnapshot(ctx context.Context, pool *pgxpool.Pool, schema string) (*Snapshot, error) {
	if err := pg.ValidateIdentifier("schema", schema); err != nil {
		return nil, err
	}
	q := fmt.Sprintf(
		`SELECT snapshot FROM %s.keystone_baselines WHERE kind = 'structure'`,
		pg.QuoteIdentifier(schema),
	)
	var raw []byte
	err := pool.QueryRow(ctx, q).Scan(&raw)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("read snapshot: %w", err)
	}
	if len(raw) == 0 {
		return nil, nil
	}
	var snap Snapshot
	if err := json.Unmarshal(raw, &snap); err != nil {
		return nil, fmt.Errorf("unmarshal snapshot: %w", err)
	}
	return &snap, nil
}

// WriteSnapshot stores the canonical JSON alongside the existing
// hash row. Called from DriftController after WriteBaseline so the
// hash and snapshot stay in sync.
func WriteSnapshot(ctx context.Context, pool *pgxpool.Pool, schema string, snap *Snapshot) error {
	if err := pg.ValidateIdentifier("schema", schema); err != nil {
		return err
	}
	enc, err := json.Marshal(snap)
	if err != nil {
		return fmt.Errorf("marshal snapshot: %w", err)
	}
	stmt := fmt.Sprintf(
		`UPDATE %s.keystone_baselines SET snapshot = $1 WHERE kind = 'structure'`,
		pg.QuoteIdentifier(schema),
	)
	if _, err := pool.Exec(ctx, stmt, enc); err != nil {
		return fmt.Errorf("write snapshot: %w", err)
	}
	return nil
}

File: rust/keystone-sdk/src/drift/snapshot_store.rs
// SPDX-License-Identifier: Apache-2.0

//! Canonical-snapshot JSON storage.
//!
//! Ported from the Go `drift` package's `snapshot_store.go`. Extends
//! `keystone_baselines` with a `snapshot JSONB` column: the DriftController
//! writes the canonical snapshot when it accepts a baseline and reads it
//! when drift is detected to compute structured findings.
//!
//! Storage uses tokio-postgres's `Json` wrapper (JSONB). PostgreSQL
//! normalises JSONB on input regardless of the producer's exact escaping,
//! so a snapshot written here is byte-identical on disk to one written by
//! the Go SDK; the round-trip reconstructs the same [`Snapshot`]. (The
//! hash-compatible formatter in [`super::marshal_snapshot`] is only needed
//! for [`super::hash`], not for JSONB storage.)

use tokio_postgres::types::Json;
use tokio_postgres::Client;

use crate::ident::{quote_identifier, validate_identifier, IdentError};

use super::Snapshot;

/// DDL adding the `snapshot JSONB` column. The single `{schema}` token is
/// substituted with the quoted schema. Indentation is a literal TAB.
const SNAPSHOT_TABLE_DDL: &str = "
ALTER TABLE {schema}.keystone_baselines
\tADD COLUMN IF NOT EXISTS snapshot JSONB;
";

fn render_ddl(schema: &str) -> String {
    SNAPSHOT_TABLE_DDL.replace("{schema}", &quote_identifier(schema))
}

/// Idempotently adds the `snapshot JSONB` column to `keystone_baselines`.
pub async fn ensure_snapshot_column(
    client: &Client,
    schema: &str,
) -> Result<(), SnapshotStoreError> {
    validate_identifier("schema", schema)?;
    client.batch_execute(&render_ddl(schema)).await?;
    Ok(())
}

/// Returns the prior snapshot JSON for `kind = 'structure'`, or `None` if no
/// row exists or the column is NULL.
pub async fn read_snapshot(
    client: &Client,
    schema: &str,
) -> Result<Option<Snapshot>, SnapshotStoreError> {
    validate_identifier("schema", schema)?;
    let q = format!(
        "SELECT snapshot FROM {}.keystone_baselines WHERE kind = 'structure'",
        quote_identifier(schema),
    );
    let rows = client.query(&q, &[]).await?;
    let Some(row) = rows.first() else {
        return Ok(None);
    };
    let snap: Option<Json<Snapshot>> = row.get(0);
    Ok(snap.map(|j| j.0))
}

/// Stores the canonical snapshot alongside the existing hash row. Called
/// after [`super::write_baseline`] so hash and snapshot stay in sync.
pub async fn write_snapshot(
    client: &Client,
    schema: &str,
    snap: &Snapshot,
) -> Result<(), SnapshotStoreError> {
    validate_identifier("schema", schema)?;
    let stmt = format!(
        "UPDATE {}.keystone_baselines SET snapshot = $1 WHERE kind = 'structure'",
        quote_identifier(schema),
    );
    client.execute(&stmt, &[&Json(snap)]).await?;
    Ok(())
}

/// Errors raised by the snapshot-store helpers.
#[derive(Debug, thiserror::Error)]
pub enum SnapshotStoreError {
    #[error(transparent)]
    Ident(#[from] IdentError),
    #[error(transparent)]
    Db(#[from] tokio_postgres::Error),
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ddl_renders_quoted_and_exact() {
        let ddl = render_ddl("public");
        let expected = "
ALTER TABLE \"public\".keystone_baselines
\tADD COLUMN IF NOT EXISTS snapshot JSONB;
";
        assert_eq!(ddl, expected);
    }
}
